daemon/k8s_watcher.go

```diff
@@ -1348,77 +1374,86 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
+				// function.
 				serverRuleCpy := serverRule.DeepCopy()
-				_, err2 = serverRuleCpy.Parse()
+				_, err2 := serverRuleCpy.Parse()
 				if err2 != nil {
 					// If we can't parse the rule then we should signalize
 					// it in the status
 					log.WithError(err2).WithField(logfields.Object, logfields.Repr(serverRuleCpy)).
 						Warn("Error parsing new CiliumNetworkPolicy rule")
 				}
 
-				cnpns := cilium_v2.CiliumNetworkPolicyNodeStatus{
-					OK: true,
-					// If the deadline was reached then not all endpoints are enforcing
-					// the given policy.
-					Enforcing:   waitForEPsErr == nil,
-					Revision:    rev,
-					LastUpdated: cilium_v2.NewTimestamp(),
-					Annotations: cnp.Annotations,
-				}
+				var err3 error
 
-				// Most important is the error while adding the CNP
-				if err != nil {
-					cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-						Error:       err.Error(),
-						OK:          false,
-						LastUpdated: cilium_v2.NewTimestamp(),
-					}
-				} else if err2 != nil {
-					cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-						Error:       err2.Error(),
-						OK:          false,
-						LastUpdated: cilium_v2.NewTimestamp(),
-					}
+				// Update the status of whether the rule is enforced on this node.
+				// If we are unable to parse the CNP retrieved from the store,
+				// or if endpoints did not reach the desired policy revision
+				// after 30 seconds, then mark the rule as not being enforced.
+				if err2 != nil {
+					err3 = updateCNPNodeStatus(serverRuleCpy, false, err2, 0, cnp.Annotations)
+				} else {
+					// If the deadline by the above context, then not all
+					// endpoints are enforcing the given policy, and
+					// waitForEpsErr will be non-nil.
+					err3 = updateCNPNodeStatus(serverRuleCpy, waitForEPsErr == nil, nil, rev, cnp.Annotations)
 				}
 
-				nodeName := node.GetName()
-				serverRuleCpy.SetPolicyStatus(nodeName, cnpns)
-				ns := k8sUtils.ExtractNamespace(&serverRuleCpy.ObjectMeta)
-
-				switch {
-				case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
-					_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).UpdateStatus(serverRuleCpy)
-				default:
-					_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(serverRuleCpy)
-				}
-				if err2 == nil {
+				if err3 == nil {
 					scopedLog.WithField("status", serverRuleCpy.Status).Debug("successfully updated with status")
 				} else {
-					return err2
+					return err3
 				}
+
 				return waitForEPsErr
 			},
 		},
 	)
 }
 
+func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing bool, err error, rev uint64, annotations map[string]string) error {
+	var (
+		cnpns cilium_v2.CiliumNetworkPolicyNodeStatus
+		err2  error
+	)
+	if err != nil {
+		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
+			Enforcing:   enforcing,
+			Error:       err.Error(),
+			OK:          false,
+			LastUpdated: cilium_v2.NewTimestamp(),
```

Comment:
Should the annotations be updated on an error case as well?

Reply:
No, the reason we want to put the Annotations in the Status of the CNP is to be able to reliably determine when a policy rule that already exists, is updated and is realized for all endpoints on the node. See the initial issue for more context: https://github.com/cilium/cilium/issues/4248

On second thought, I guess we can add the annotations if `OK` is `false`, as a CNP being `OK` is a prerequisite for it being enforced on a given node. Is this what you are implying?

Reply:
No, I believe that was a simple flag to acknowledge if there was something not OK.

Reply:
I think the Annotations can be used to show when the rule has been at least added to Cilium; it is up to those who utilize the CNP Status to check the error and whether it is OK, so I will unconditionally update the Annotations.